### DISCUSSION TITLE
Thermodynamics

### DISCUSSION THREAD

**Discussion by dvd101x** [mgreminger/EngineeringPaper.xyz#204] [Ideas]
Hi,

I was wondering if [CoolProp](http://www.coolprop.org/) would be desirable for this project in cases when thermodynamics are involved.

Here is a reference
https://observablehq.com/@dvd101x/vapor-compression-cycle

**Comment by mgreminger** [comment]
Yes, there are intentions to eventually add some thermodynamics specific functions. I was considering the [thermo](https://github.com/CalebBell/thermo) library but the CoolProp library you point out looks promising. A Pyodide package would need to be built since it's not a pure Python library. Looking at your observable notebook, it has been compiled to wasm at some point for use in JavaScript so it should be possible to build a Pyodide package for it.

**Reply by dvd101x** [reply]
Thanks for the response, I will take a better look at thermo, at some point I was reviewing also [pyrmoat](http://pyromat.org/)

I have something working in pyodide that imports the compiled wasm + javascript. I will include a code snippet at the end of this comment if you would like to take a look.

They are also working on adding Coolprop to pyodide.
https://github.com/pyodide/pyodide/pull/4028

```html
<!doctype html>
<html>

<head>
    <script src="https://cdn.jsdelivr.net/npm/coolporp@1.0.0/coolprop.js"></script>
    <script src="https://cdn.jsdelivr.net/pyodide/dev/full/pyodide.js"></script>
</head>

<body>
    Pyodide test page <br>
    Open your browser console to see Pyodide output
    <script type="text/javascript">
        async function main() {
            let pyodide = await loadPyodide();
            console.log(pyodide.runPython(`
            import sys
            sys.version
            `));
            pyodide.runPython(`
            import js
            PropsSI = js.Module.PropsSI
            print(PropsSI('D', 'T', 298.15, 'P', 101325, 'Nitrogen'))
            `)
        }
        main();
    </script>
</body>

</html>
```

**Reply by dvd101x** [reply]
Hi, just an update

Pyodide merged the PR to include coolprop.

I did a tweak to the code I included on the last comment the other major functions of coolprop.

```html
<!doctype html>
<html>

<head>
    <script src="https://cdn.jsdelivr.net/npm/coolporp@1.0.0/coolprop.js"></script>
    <script src="https://cdn.jsdelivr.net/pyodide/dev/full/pyodide.js"></script>
</head>

<body>
    Pyodide test page <br>
    Open your browser console to see Pyodide output
    <script type="text/javascript">
        async function main() {
            let pyodide = await loadPyodide();
            console.log(pyodide.runPython(`
            import sys
            sys.version
            `));
            pyodide.runPython(`
            import js
            PropsSI = js.Module.PropsSI
            HAPropsSI = js.Module.HAPropsSI

            phases = ["liquid", "supercritical", "supercritical_gas", "supercritical_liquid", "critical_point", "gas", "twophase", "unknown", "not_imposed"]

            def PhaseSI(prop1, val1, prop2, val2, fluid):
                return phases[PropsSI("Phase", prop1, val1, prop2, val2, fluid)]

            # Density of Nitrogen at 298.15 K and 101325 Pa in kg/m^3
            print(PropsSI('D', 'T', 298.15, 'P', 101325, 'Nitrogen'))

            # Enthalpy (J per kg dry air) as a function of temperature, pressure,
            #    and relative humidity at dry bulb temperature T of 25C, pressure
            #    P of one atmosphere, relative humidity R of 50%
            print(HAPropsSI("H", "T", 298.15, "P", 101325, "R", 0.5))

            # Phase of water at 101325 Pa and Quality 0%
            print(PhaseSI("P", 101325, "Q", 0, "Water"))
            `)
        }
        main();
    </script>
</body>

</html>
```

**Reply by dvd101x** [reply]
Hi,

Just to let you know that pyodide now support coolprop. So that workaround I mentioned is no longer necessary.

https://pyodide.org/en/stable/usage/packages-in-pyodide.html#packages-in-pyodide

**Reply by dvd101x** [reply]
Hi, just found that thermo uses coolprop (and others).

https://github.com/CalebBell/thermo/blob/42649a952bf39678b33de6fb2ba2a34a72f2ddb7/environment.yml

I don't want to bother you with this, as this project is having great progress. Just wanted to share this maybe for the future.

**Reply by mgreminger** [reply]
Thanks for the updates. Yes, it looks like coolprop is the way to go, especially since it has been added to Pyodide. I'm assuming the PropsSI() and HAPropsSI() functions would get most of what you would need. The implementation is going to be fairly involved in order to handle the units properly for all of the possible inputs and outputs. Also, strings are not currently supported as inputs to functions so some work would be required in order to support specifying the inputs, output, and fluid for each function call.

**Reply by dvd101x** [reply]
Yes, you are right. It's mostly PropsSi(), HAPropsSI() and Phase(). Regarding the units, as you mentioned is not that straight forward, I have an implementation in javascript that is basically wrapping those functions to handle units with mathjs to get the input and outputs as the CoolProp docs state. Regarding the strings, it would be a significant change, maybe part of it could be accomplished by a table or dropdown, but there are some special cases where they allow for very specific strings to do mixtures or integrations.

**Reply by mgreminger** [reply]
Looks like the issue of a [CoolProp exception crashing Pyodide](https://github.com/CoolProp/CoolProp/discussions/2302#discussioncomment-7333820) has not been resolved (I just tried it on the latest Pyodide REPL using Pyodide 0.25.0, see below). We won't be able to integrate CoolProp into EngineeringPaper.xyz until this issue is resolved. 

```pycon
Welcome to the Pyodide terminal emulator 🐍
Python 3.11.3 (main, Jan 18 2024 19:07:12) on WebAssembly/Emscripten
Type "help", "copyright", "credits" or "license" for more information.
>>> from CoolProp.CoolProp import PropsSI
>>> PropsSI('T','P',101325,'Q',0,'Water')
373.1242958476844
>>> PropsSI('T','P',101325,'Q',0,'Waterz')
Pyodide has suffered a fatal error. Please report this to the Pyodide maintainers.
The cause of the fatal error was:
CppException CoolProp::CoolPropError<(CoolProp::CoolPropBaseError::ErrCode)4>: key [Waterz] was not found in string_to_index_map in JSONFluidLibrary
Look in the browser console for more details.
```

**Reply by dvd101x** [reply]
Thanks for looking into this.

It doesn't appear this issue has been reported in the Pyodide project.

A workaround could be to do it with the javascript wrapper. But maybe it's too cumbersome or not inline with this project.

I will report the issue in the Pyodide project
If you would like to test the javascript alternative I'm open to collaborate, but of course it's your call

**Reply by dvd101x** [reply]
Created the issue at https://github.com/pyodide/pyodide/issues/4451

**Reply by mgreminger** [reply]
Thanks for reporting the issue to Pyodide. It looks like there's a way to fix this based on the issue discussion. My preferred path would be to used the Pyodide package directly. I'm a little concerned about the stability and performance of the JavaScript workaround.

**Reply by dvd101x** [reply]
Hi, just to let you know there has been some progress with the pyodide [PR](https://github.com/pyodide/pyodide/pull/4455) to fix this issue.

**Reply by mgreminger** [reply]
I'm currently working on implementing the CoolProps integration (#257). Thanks for working on getting the error handling fixed with CoolProp in Pyodide and thanks for suggesting CoolProp as the library to use for thermo fluid properties. If everything goes as planned, I should be able to roll out the initial implementation of the feature in a week or two.

**Reply by dvd101x** [reply]
Thanks for the update!

I've been following the progress and It looks very cool!

I noticed some interest from the CoolProp guys in here.
https://github.com/CoolProp/CoolProp/discussions/2358

**Reply by mgreminger** [reply]
Yes, I can confirm that the 0.26.0 official release of Pyodide does include the CroolProp exception handling fix. I updated the #257 feature branch a few days ago to include the fix. The speed of the Pyodide version of CoolProp looks quite impressive from the benchmarks that you ran.

**Reply by mgreminger** [reply]
This feature is now released! You can check out the tutorial here: https://engineeringpaper.xyz/h2x75xbLYjEgg7uWgXcHvz

**Reply by dvd101x** [reply]
Oh wow, that's awesome!

I did read the tutorial and made some tests on my own. It's impressive and I think it opens a world of possibilities!

**Reply by mgreminger** [reply]
Just a heads up that I'm working on implementing parametric plotting in EP (#259), which will make it possible to create P-H phase diagrams like in your observable example above. It's something I've been meaning to add for a while.

**Reply by dvd101x** [reply]
Oh, that's very cool.

I will try to clean up my observable example so it's more readable.

**Reply by mgreminger** [reply]
The parametric plotting feature has been rolled out now. I've recreated a version of your P-H diagram here: https://engineeringpaper.xyz/nvNHofmnhULuyThoRzthe8

**Reply by mgreminger** [reply]
Updated it to include the COP and Qc calculations from your observable example: https://engineeringpaper.xyz/mxwD7wdVwGGc43kVKPWgvy
